### PR TITLE
padded string improved

### PR DESCRIPTION
**Pull request by DBJDBJ**
Checking arguments in debug builds
Added constructor with std::string_view argument
`std::string` argument now passed by `const &` not by value 
`allocate_padded_buffer()` moved in here, also with added length argument sanity check in debug builds

### CODE REVIEW

File: include/simdjson/padded_string.h [hunk 1]
```diff
@@ -10,34 +10,92 @@ namespace simdjson {
...
+    else {
+      errno = EINVAL;
+      perror("simdjson::padded_string() data argument is null");
+      // what to do here? exit() ?
+    }
+#endif
   }
-  padded_string(std::string s) noexcept
+
+  // note: do not have std::string arguments passed by value
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Please rephrase... like " do not pass std::string arguments by value"

File: include/simdjson/padded_string.h [hunk 2]
```diff
@@ -10,34 +10,92 @@ namespace simdjson {
...
+  padded_string(std::string const &s) noexcept
```
Comment (lemire) [inline include/simdjson/padded_string.h]: I think that elsewhere in the library we put the const qualifiers before the type. Unless you have a good reason to break the style, I recommend moving the const qualifier before std::string.

File: include/simdjson/padded_string.h [hunk 3]
```diff
@@ -60,21 +118,31 @@ class padded_string {
...
+  char *data() const noexcept { return data_ptr; }
 
 private:
   padded_string &operator=(const padded_string &o) = delete;
   padded_string(const padded_string &o) = delete;
 
-  size_t viable_size;
-  char *data_ptr;
-};
+  size_t viable_size{};
```
Comment (lemire) [inline include/simdjson/padded_string.h]: What is the purpose of the empty braces?

File: include/simdjson/padded_string.h [hunk 4]
```diff
@@ -60,21 +118,31 @@ class padded_string {
...
+    if (this->data_ptr != nullptr)
+#endif
+    {
+      aligned_free_char(data_ptr);
+      this->data_ptr = nullptr;
+    }
+  }
 
-  size_t size() const { return viable_size; }
+  size_t size() const noexcept { return viable_size; }
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Does noexcept serve a purpose here?

File: include/simdjson/padded_string.h [hunk 5]
```diff
@@ -60,21 +118,31 @@ class padded_string {
...
 
-  size_t length() const { return viable_size; }
+  size_t length() const noexcept { return viable_size; }
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Does noexcept serve a purpose here?

File: include/simdjson/padded_string.h [hunk 6]
```diff
@@ -60,21 +118,31 @@ class padded_string {
...
 
-  char *data() const { return data_ptr; }
+  char *data() const noexcept { return data_ptr; }
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Does noexcept serve a purpose here?

File: include/simdjson/padded_string.h [hunk 7]
```diff
@@ -60,21 +118,31 @@ class padded_string {
     o.viable_size = tmp_viable_size;
   }
 
-  ~padded_string() { aligned_free_char(data_ptr); }
+  ~padded_string() {
+#ifndef NDEBUG
+    if (this->data_ptr != nullptr)
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Is this needed? Calling free on a null pointer ought to be safe.

File: include/simdjson/padded_string.h [hunk 8]
```diff
@@ -10,34 +10,92 @@ namespace simdjson {
...
+      errno = EINVAL;
+      perror("simdjson::padded_string() std::string argument is empty");
+      // what to do here? exit() ?
+    }
+#endif
   }
+
+  padded_string(std::string_view sv_) noexcept
+      : viable_size(sv_.size()), data_ptr(allocate_padded_buffer(sv_.size())) {
+    if (data_ptr != nullptr) {
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Formally speaking, we should check that sv_.data() is not null.

File: include/simdjson/padded_string.h [hunk 9]
```diff
@@ -10,34 +10,92 @@ namespace simdjson {
...
       : viable_size(s.size()), data_ptr(allocate_padded_buffer(s.size())) {
     if (data_ptr != nullptr) {
       memcpy(data_ptr, s.data(), s.size());
       data_ptr[s.size()] = '\0'; // easier when you need a c_str
     }
+#ifndef NDEBUG
+    else {
+      errno = EINVAL;
+      perror("simdjson::padded_string() std::string argument is empty");
+      // what to do here? exit() ?
```
Comment (lemire) [inline include/simdjson/padded_string.h]: No. I think that if memory alloc fails, we are left with a padded_string that has a null pointer. In such cases, we should just set viable_size to zero.

File: include/simdjson/padded_string.h [hunk 10]
```diff
@@ -10,34 +10,92 @@ namespace simdjson {
...
+      : viable_size(sv_.size()), data_ptr(allocate_padded_buffer(sv_.size())) {
+    if (data_ptr != nullptr) {
+      memcpy(data_ptr, sv_.data(), sv_.size());
+      data_ptr[sv_.size()] = '\0'; // easier when you need a c_str
+    }
+#ifndef NDEBUG
+    else {
+      errno = EINVAL;
+      perror("simdjson::padded_string() std::string_view argument is empty");
+      // what to do here? exit() ?
```
Comment (lemire) [inline include/simdjson/padded_string.h]: No. In such cases, we should just set viable_size to zero.

File: include/simdjson/padded_string.h [hunk 11]
```diff
@@ -10,34 +10,92 @@ namespace simdjson {
...
+    perror("simdjson::allocate_padded_buffer() length argument is less than 1");
+    return nullptr;
+  }
+#endif // NDEBUG
+
+  // we could do a simple malloc
+  // return (char *) malloc(length + SIMDJSON_PADDING);
+  // However, we might as well align to cache lines...
+  size_t totalpaddedlength = length + SIMDJSON_PADDING;
+  char *padded_buffer = aligned_malloc_char(64, totalpaddedlength);
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Technically, we should check that the allocation worked before continuing.

File: include/simdjson/padded_string.h [hunk 12]
```diff
@@ -10,34 +10,92 @@ namespace simdjson {
...
+  memset(padded_buffer + length, 0, totalpaddedlength - length);
+  return padded_buffer;
+} // allocate_padded_buffer
 
 // Simple string with padded allocation.
 // We deliberately forbid copies, users should rely on swap or move
 // constructors.
-class padded_string {
-public:
+struct padded_string final {
```
Comment (lemire) [inline include/simdjson/padded_string.h]: The change to struct is fine, but why "final"?

File: include/simdjson/padded_string.h [hunk 13]
```diff
@@ -10,34 +10,92 @@ namespace simdjson {
...
+      perror("simdjson::padded_string() length argument is less than 1");
+      // what to do here? exit() ?
+    }
+#endif
   }
+
   explicit padded_string(char *data, size_t length) noexcept
       : viable_size(length), data_ptr(allocate_padded_buffer(length)) {
     if (data_ptr != nullptr) {
       memcpy(data_ptr, data, length);
```
Comment (lemire) [inline include/simdjson/padded_string.h]: Technically, we should check that "data" is not null here.

**Comment by DBJDBJ** [comment]
ammended

**Comment by lemire** [comment]
@DBJDBJ The tests fail because we sometimes parse an empty string. The problem is the new test you have added in `allocate_padded_buffer` to forbid 0 byte allocations. I don't think we want a 0-byte allocation to crash the program.

**Comment by lemire** [comment]
This PR should not change the files windows/dirent_portable.h and windows/toni_ronnko_dirent.h. I don't understand these changes.

**Comment by DBJDBJ** [comment]
Neither do I? there are no changes in there... I am even in a different branch.  And Visual Studio now has no idea about git for this solution.  Let me look into that.

ps: you want to parse an empty string? OK. I assume as some edge case ...

**Comment by lemire** [comment]
> you want to parse an empty string? OK. I assume as some edge case ...

Well, we don't want the system to crash if there is an empty string.

**Comment by DBJDBJ** [comment]
@lemire What do you want to happen on `allocate_padded_buffer(0)`?
I assume just make and return 
```cpp
size_t totalpaddedlength = 0 /*length*/ + SIMDJSON_PADDING;
  char *padded_buffer = aligned_malloc_char(64, totalpaddedlength);
```

**Comment by lemire** [comment]
Something like that, I suppose, yes.

**Comment by DBJDBJ** [comment]
ok lets try a new PR

**Pull request closed without merging**